nri_device_injector/nri_device_injector.go

```diff
@@ -0,0 +1,191 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"golang.org/x/sys/unix"
+	"sigs.k8s.io/yaml"
+
+	"github.com/containerd/nri/pkg/api"
+	"github.com/containerd/nri/pkg/stub"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	deviceKeyPrefix = "devices.gke.io"
+	blockDevice     = "b"
+	charDevice      = "c"
+	fifoDevice      = "p"
+	pluginName      = "device_injector_nri"
+	pluginIdx       = "10"
+)
+
+type device struct {
+	Path     string `json:"path"`
+	Type     string `json:"type"`
+	Major    int64  `json:"major"`
+	Minor    int64  `json:"minor"`
+	FileMode uint32 `json:"file_mode"`
+	UID      uint32 `json:"uid"`
+	GID      uint32 `json:"gid"`
+}
+
+type plugin struct {
+	stub stub.Stub
+}
+
+func main() {
+	var (
+		opts []stub.Option
+		err  error
+	)
+
+	opts = append(opts, stub.WithPluginName(pluginName))
+	opts = append(opts, stub.WithPluginIdx(pluginIdx))
+
+	p := &plugin{}
+
+	if p.stub, err = stub.New(p, append(opts, stub.WithOnClose(p.onClose))...); err != nil {
+		log.Errorf("Failed to create plugin stub: %v", err)
+		os.Exit(1)
+	}
+
+	err = p.stub.Run(context.Background())
+	if err != nil {
+		log.Errorf("plugin exited with error %v", err)
+		os.Exit(1)
+	}
+}
+
+func (p *plugin) onClose() {
+	log.Info("NRI connection closed")
+}
+
+// CreateContainer handles CreateContainer requests relayed to the plugin by containerd NRI.
+// The plugin makes adjustment on containers annotated to have devices injected.
+func (p *plugin) CreateContainer(_ context.Context, pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+	var (
+		ctrName string
+		devices []device
+		err     error
+	)
+
+	ctrName = container.Name
+	log.Infof("Started CreateContainer on container: %v, pod: %v, namespace: %v", ctrName, pod.Name, pod.Namespace)
+	if pod != nil {
+		ctrName = pod.Name + "/" + ctrName
+	}
+
+	adjust := &api.ContainerAdjustment{}
+
+	devices, err = getDevices(container.Name, pod.Annotations)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(devices) == 0 {
+		log.Infof("Container %s: no devices annotated...", ctrName)
+	} else {
+		for _, d := range devices {
+			log.Infof("Container %s: annotated device %q...", ctrName, d.Path)
+			deviceNRI, err := d.toNRIDevice()
+			if err != nil {
+				return nil, nil, err
+			}
+			adjust.AddDevice(deviceNRI)
+			log.Infof("Container %s: injected device %q...", ctrName, d.Path)
+		}
+	}
+
+	return adjust, nil, nil
+}
+
+// getDevices returns parsed devices from pod annotations of device injections.
+func getDevices(ctrName string, podAnnotations map[string]string) ([]device, error) {
+	var (
+		key        string
+		annotation []byte
+		devices    []device
+	)
+
+	for _, key = range []string{
+		deviceKeyPrefix + "/container." + ctrName,
+		deviceKeyPrefix + "/pod",
+		deviceKeyPrefix,
+	} {
+		if value, ok := podAnnotations[key]; ok {
+			annotation = []byte(value)
+			break
+		}
+	}
```

Comment:
can there be multiple annotations and we only pick them by this order of preference, first container, second pod , ... ? 
if that is the logic can we add a comment explaining it?

Reply:
There can be multiple annotations, and here I'm making container level and pod level specs exclusive to each other to reduce complexity. I added some comments to explain this.

Reply:
Do we have a need for pod-level injection?  If we're really only looking at container-level for our immediate use-cases, it'd be simpler to remove the pod-level logic for now and add it back if we ever have a future need.

Reply:
@samuelkarp Makes sense, I've removed the pod level injection and only allow container level injection. Thanks!